bluetooth: sdp: fix the corner cases for the sdp discover

MarkWangChinese opened the pull request:
1. bt_sdp_discover may be called in two tasks at the same time: (1) If the 2 calls are for 2 acl conns,
the same session may be new allocated to two acl conns without protection. (2) If the 2 calls are for the same acl conn, the same session should be used.
2. fix that bt_sdp_discover may use one session that is disconnecting.
3. If the sdp l2cap connecting fail, the pending reqs need callback.

MarkWangChinese commented:
Run the two sdp tests, the result pass as follow:
INFO    - 1/1 mimxrt1170_evk@B/mimxrt1176/cm7 bluetooth.classic.sdp.client                       [32mPASSED[39m (device 64.740s <zephyr>)
INFO    - 1 test scenarios (1 configurations) selected, 0 configurations filtered (0 by static filter, 0 at runtime).
INFO    - [32m1 of 1[39m executed test configurations passed (100.00%), 0 built (not run), 0 failed, 0 errored, with no warnings in 179.40 seconds.
INFO    - 12 of 12 executed test cases passed (100.00%) on 1 out of total 991 platforms (0.10%).
INFO    - [32m1[39m test configurations executed on platforms, [36m0[39m test configurations were only built.


INFO    - 1/1 mimxrt1170_evk@B/mimxrt1176/cm7 bluetooth.classic.sdp.server                       PASSED (device 21.758s <zephyr>)
INFO    - 1 test scenarios (1 configurations) selected, 0 configurations filtered (0 by static filter, 0 at runtime).
INFO    - 1 of 1 executed test configurations passed (100.00%), 0 built (not run), 0 failed, 0 errored, with no warnings in 30.35 seconds.
INFO    - 2 of 2 executed test cases passed (100.00%) on 1 out of total 991 platforms (0.10%).
INFO    - 1 test configurations executed on platforms, 0 test configurations were only built.

jhedberg commented on subsys/bluetooth/host/classic/sdp.c:
```diff
@@ -1450,6 +1456,11 @@ void bt_sdp_init(void)
 		LOG_ERR("L2CAP server registration failed with error %d", res);
 	}
+
+	for (int i = 0; i < sizeof(bt_sdp_client_pool) / sizeof(bt_sdp_client_pool[0]); i++) {
```
`ARRAY_FOR_EACH(bt_sdp_client_pool, i) {`

lylezhu2012 commented on subsys/bluetooth/host/classic/sdp.c:
```diff
@@ -2252,62 +2294,81 @@ static struct bt_sdp_client *sdp_client_new_session(struct bt_conn *conn)
+	 */
+	k_sem_take(&sdp_sem_lock, K_FOREVER);
 	for (i = 0; i < ARRAY_SIZE(bt_sdp_client_pool); i++) {
 		if (bt_sdp_client_pool[i].chan.chan.conn == conn) {
-			return &bt_sdp_client_pool[i];
+			k_sem_take(&bt_sdp_client_pool[i].sem_lock, K_FOREVER);
+			if (bt_sdp_client_pool[i].valid) {
+				sys_slist_append(&bt_sdp_client_pool[i].reqs,
+						 (sys_snode_t *)&params->_node);
+				k_sem_give(&bt_sdp_client_pool[i].sem_lock);
+				return 0;
+			}
+			k_sem_give(&bt_sdp_client_pool[i].sem_lock);
 		}
 	}
```
I think we can use index of ACL conn to get the sdp client bt_sdp_client_pool[<acl index>]. And then use the atomic flags to flag the status.

jhedberg replied:
@lylezhu2012 that sounds like a good idea. This is why we have `bt_conn_index()`

MarkWangChinese replied:
OK, let me try it. Then one acl conn match with one session of `bt_sdp_client_pool`. If the `session` is in disconnecting, the  bt_sdp_discover needs to wait the disconnecting finish, then create the sdp l2cap connection again with the same session.

jhedberg commented on subsys/bluetooth/host/classic/sdp.c:
```diff
@@ -2230,84 +2277,134 @@ static void sdp_client_disconnected(struct bt_l2cap_chan *chan)
-		session->chan.chan.ops = &sdp_client_chan_ops;
-		session->chan.chan.conn = conn;
-		session->chan.rx.mtu = SDP_CLIENT_MTU;
+		conn = bt_conn_lookup_index(session - &bt_sdp_client_pool[0]);
```
Is that actually desirable? Should it be changed? (so that the clearing happens right after the callback)

MarkWangChinese replied:
From my understand. I think it is right behavior because the `chan` should not be touched again after the `released` callback. In the `released` callback, the `chan` is released and can be used for new l2cap connection.

jhedberg replied:
Ok, fair enough. However, please use the existing `ARRAY_INDEX()` macro instead of doing the subtraction yourself.

MarkWangChinese commented on subsys/bluetooth/host/classic/sdp.c:
```diff
@@ -96,6 +107,8 @@ struct bt_sdp_client {
 	/* Received data length */
 	uint32_t                             recv_len;
+	/* client state */
+	uint8_t                              state;
```
Fixed.

jhedberg commented on subsys/bluetooth/host/classic/sdp.c:
```diff
@@ -2230,84 +2277,134 @@ static void sdp_client_disconnected(struct bt_l2cap_chan *chan)
 	 */
 	(void)memset(&session->reqs, 0,
-		     sizeof(*session) - sizeof(session->chan));
+		     sizeof(*session) - offsetof(struct bt_sdp_client, reqs));
```
The code comment above doesn't seem valid after the change. In two other places you're also doing something different: 
```c
(void)memset(&session->chan, 0,
		sizeof(*session) - offsetof(struct bt_sdp_client, chan));
```

This playing around with offsets, pointers and memset will quickly become a potential source of bugs in addition to also being hard to follow. I'd recommend rethinking this, perhaps add some well documented helper functions, or do the resetting of the object using some more explicit method than memset.

MarkWangChinese replied:
Updated with explicit function, please help to review again.

jhedberg commented on subsys/bluetooth/host/classic/sdp.c:
```diff
@@ -2206,108 +2234,193 @@ static void sdp_client_connected(struct bt_l2cap_chan *chan)
+	 * chan - it is still used before released callback.
+	 * reqs_next - the pending reqs in the disconnecting phase.
+	 */
+	(void)memset(&session->reqs, 0,
+		sizeof(*session) - offsetof(struct bt_sdp_client, reqs));
+}
+
+static void sdp_client_clean_after_release(struct bt_sdp_client *session)
+{
+	/*
+	 * keep the follow fields:
+	 * sem_lock - global, never clean it.
+	 */
+	/* session->state is set as SDP_CLIENT_RELEASED by memset */
+	(void)memset(&session->state, 0,
+		     sizeof(*session) - offsetof(struct bt_sdp_client, state));
```
This still looks prone to bugs. E.g. if `session->recv_buf` is set it should be freed with `net_buf_unref()`. I'd propose just doing each member individually, and you can probably reuse one of these "clean()" functions from the other, i.e. `after_release()` can probably call `after_disconnect()` first to avoid duplicating too many lines of code.

MarkWangChinese replied:
agree, updated.

MarkWangChinese commented on subsys/bluetooth/host/classic/sdp.c:
```diff
@@ -2206,108 +2234,203 @@ static void sdp_client_connected(struct bt_l2cap_chan *chan)
+{
+	/*
+	 * keep the follow fields:
+	 * sem_lock - global, never clean it.
```
Right. Updated for review.

jhedberg commented:
@MarkWangChinese please split this into multiple commits, if possible. One commit per fix. Whenever you notice that you have to enumerate multiple independent things in the commit message or PR description, it's a good indication that things need to be split up into multiple commits.

MarkWangChinese committed:
bluetooth: sdp: callback the pending reqs when sdp l2cap connecting fail

If the sdp l2cap connecting fail, the pending reqs need callback.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese committed:
bluetooth: sdp: fix the corner cases for the sdp discover

1. bt_sdp_discover may be called in two tasks at the same time:
(1) If the 2 calls are for 2 acl conns,
the same session may be new allocated to two acl conns.
(2) If the 2 calls are for the same acl conn, the same session should be
used.
Use the bt_conn_index to map acl conn with session to fix it.
2. fix that bt_sdp_discover may use one session that is disconnecting,
put req to reqs_next and process it after disconnected.
3. remove memset of session, reset each members individually.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>
(cherry picked from commit dcbe5d7dc5d72ce0d2915cf2b291cc84a916a536)

MarkWangChinese committed:
bluetooth: sdp: cast from 'const x *' to 'struct x *' drops const qualifier

SonarQube reports: cast from 'const struct _snode *' to 'struct _snode *'
drops const qualifier. Fix it by removing the const.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>
(cherry picked from commit 7982ee0df13b7f9e54595f799e32f38c464b0f9e)

MarkWangChinese commented on subsys/bluetooth/host/classic/sdp.c:
```diff
@@ -2206,108 +2234,203 @@ static void sdp_client_connected(struct bt_l2cap_chan *chan)
+	/* put in `reqs_next` for next round after disconnected */
+	if (session->state == SDP_CLIENT_DISCONNECTING) {
+		sys_slist_append(&session->reqs_next,
+			(sys_snode_t *)&params->_node);
```
Fixed

MarkWangChinese commented on subsys/bluetooth/host/classic/sdp.c:
```diff
@@ -2206,108 +2234,203 @@ static void sdp_client_connected(struct bt_l2cap_chan *chan)
+	 * sem_lock - it is always valid to protect the session, never clean it after bt_sdp_init.
+	 */
+	session->state = SDP_CLIENT_RELEASED;
+	memset(&session->chan, 0, sizeof(session->chan));
```
removed the line `memset(&session->chan, 0, sizeof(session->chan));`

MarkWangChinese commented on subsys/bluetooth/host/classic/sdp.c:
```diff
@@ -2206,108 +2234,203 @@ static void sdp_client_connected(struct bt_l2cap_chan *chan)
 	 * connect to remote SDP endpoint.
 	 */
-	return sdp_client_new_session(conn);
+	sys_slist_init(&session->reqs);
```
no problem.

MarkWangChinese commented:
> Btw, the SonarQube issues look valid to me, i.e. we shouldn't have an API with a `const` parameter which is actually written to in the function.

@jhedberg @lylezhu2012 I added one commit to fix it, please help to review it again. Thanks.

lylezhu2012 commented:
It is not reasonable to use the same subject for different commits.

jhedberg commented on subsys/bluetooth/host/classic/sdp.c:
```diff
@@ -80,8 +80,19 @@ NET_BUF_POOL_FIXED_DEFINE(sdp_pool, CONFIG_BT_MAX_CONN, BT_L2CAP_BUF_SIZE(SDP_MT
 
 #define SDP_CLIENT_MTU 64
 
+enum __packed sdp_client_state {
```
What's the reason you're using `__packed` here? It's not benefitting you at all in the `struct bt_sdp_client` definition (since there'll anyway be padding), so the end result is actually larger code size without any benefit.

jhedberg replied:
You *can* use `__packed` when it makes sense (e.g. smaller memory footprint or because the type is required to be of a certain size due to a protocol specification), but I'm not aware of any style or rule to always use it.
